### PR TITLE
New target type tcppoll: implemented by PollSocketStream

### PR DESCRIPTION
**Pull request by davidjsherman**
Add new type of target, `tcppoll:`, implemented by `PollStream`, a socket stream that can only be polled but not read nor written.

Also fixed a wrong variable name introduced in 0686c134a48c858cc4aff86bb7c9c4827f0c9b75.

### CODE REVIEW

File: dashel/dashel-win32.cpp [hunk 1]
```diff
@@ -311,7 +311,7 @@ namespace Dashel
 			if (bindAddress.port == 0)
 			{
 				int addrSize(sizeof(addr));
-				if (getsockname(fd, (struct sockaddr *)&addr, &addrSize) != 0)
+				if (getsockname(sock, (struct sockaddr *)&addr, &addrSize) != 0)
```
Comment (davidjsherman) [inline dashel/dashel-win32.cpp]: In 0686c134a48c858cc4aff86bb7c9c4827f0c9b75 the variable name from dashel-posix.cpp was used here, breaking compilation

**Comment by stephanemagnenat** [comment]
Why not inheriting `PollSocketStream` from `SelectableStream`? It would avoid having an unused `recvBuffer` member in it.

**Comment by stephanemagnenat** [comment]
Also, shouldn't this protocol be named `tcppoll`?

File: dashel/dashel-posix.cpp [hunk 2]
```diff
@@ -554,6 +554,24 @@ namespace Dashel
 		}
 	};
 	
+	//! Poll socket, used for polling only. A target "poll:sock=N" will call Hub::incomingData(stream)
+	//! exactly once when the socket N is polled with POLLIN in Hub::step.
```
Comment (stephanemagnenat) [inline dashel/dashel-posix.cpp]: I would also explicitly state that this stream never actually reads or write data, as it aims to be used as an adapter for program reading/writing data by themselves.

**Comment by stephanemagnenat** [comment]
What about Windows, in this PR you implemented `PollSocketStream` for POSIX only?

**Comment by davidjsherman** [comment]
1. I wanted the `fd = target.get<int>("sock")` but I suppose it could be duplicated here
1. I don't actually know what it does if a TCP host is given instead of an existing socket
1. OK will improve the documentation

**Comment by stephanemagnenat** [comment]
1. It's a good point, I think the solution would be to factorize the content of the constructor of SocketStream in a support function. One solution would be multiple inheritance, but I think it's too heavy, so we can have a free floating function to which we pass the necessary contextual variables, by reference if necessary.
2. Dashel will open the connection, then only provide poll notifications, which seems correct.
3. Thanks :-)

**Comment by davidjsherman** [comment]
Target type `tcppoll` implemented in 31896634e1d49db687209d0de02a52cce92e432c by **PollStream**, derived from SelectableStream in POSIX and from WaitableStream in Win32. Refactored socket assignment in static function `GetOrCreateSocket`.

When a Pollstream socket responds to a poll, Hub calls `PollStream::incomingData` exactly once. The edge trigger for PollStream in POSIX is `bool edgeTrigger`. The edge trigger in Win32 relies on a new callback method in WaitableStream
```
virtual void notifyIncomingData(Hub *srv, EvType& t) { }
```
that is called by `Dashel::step`, and is needed by PollStream to reset the trigger.

**Comment by stephanemagnenat** [comment]
Beside the naming of the static function, this looks good to merge for me. There are two additions we should do:

1. We should document `tcppoll` in `dashel.h` at line 96, which is the list of the protocols in the autodoc. We should also details the options for `tcppoll` at line 112.
2. We should bump the medium version number to 3 (linus 137 and 139 of `Dashel.h`), meaning that Dashel will become version 1.3.0. This will allow us to test this number in the Aseba CMake file.

**Comment by stephanemagnenat** [comment]
One more things, do not forget to add yourself to `readme.txt`, which has the author list (which in an ideal world should be in its own file...).

**Comment by stephanemagnenat** [comment]
Looks good! If it's good on your side I'll merge and release 1.3.0

**Comment by davidjsherman** [comment]
Wait, does it matter that `sock=` is last in the order of the parameters, but in the logic of `tcppoll:` (and `tcp:`) it takes priority over `host=` and `port=`? I think the Dashel syntax lets all of the following targets all refer to local socket 15, but currently only the ones with an explicit `sock=` work.

1. `tcppoll:15`
1. `tcppoll:sock=15`
1. `tcppoll:www.google.com;80;15`
1. `tcppoll:www.google.com;15;port=80`
1. `tcppoll:sock=15;port=33333`
1. `tcppoll:sock=15;localhost;33333`

**Comment by stephanemagnenat** [comment]
I agree that the current behaviour, which you identified correctly, is a bit strange.

Note that 1. means connect to host "15".  Indeed out of these 6 lines only 3, 5 and 6 work as intended, as `Dashel::ParameterSet` does not do a multi-pass analysis of the line to first remove the named parameters: unnamed parameters will be assigned to the name corresponding to their position in the string, regardless if there was a named parameter before that is normally after. I agree it is limiting and unnatural, but I'm not sure it is a priority to change that.

**Comment by stephanemagnenat** [comment]
That said, we could document that if the `sock` option is used, it is forbidden to specify `host` and `port`.

**Comment by stephanemagnenat** [comment]
One more thing to do before merging: we should increase `LIB_VERSION_MAJOR` and set `LIB_VERSION_MINOR` to 0 in the `CMakeLists.txt`.

**Comment by stephanemagnenat** [comment]
Should I merge?

**Comment by davidjsherman** [comment]
OK for me

**Comment by davidjsherman** [comment]
Wait, one more problem. The base class destructor, `SelectableStream::~SelectableStream`, will close the file descriptor. But that should be the responsibility of the code that opened the socket, `PollStream`, who is just watching the file descriptor.

**Comment by davidjsherman** [comment]
If a socket file descriptor given to `PollStream` or `SocketStream` is closed on destruction, that means that `sock=N` transfers ownership. That wasn't the intention of `PollStream`: the idea was to simply watch the socket.

Indeed transferring ownership would complicate the semantics, since in a client derived Hub class it would be necessary to 1) stop writing, 2) shut down the Hub (thus closing the stream), 3) *not* close the socket, even though the client allocated it, because it would already have been closed by the Hub.

Consequently, in 82a309258502c34b7eab18bde742d873df529753 `PollStream` doesn't try to close the socket., because it is not the owner.

**Comment by stephanemagnenat** [comment]
I totally agree about not transferring ownership. Now the question is whether `SocketStream` should implement the same logic?

**Pull request merged**